### PR TITLE
Don't output parameter initializer statement if default value is nil

### LINKED ISSUE #1422: [TSTL transformation] Default parameters with 'null' and 'undefined' get compiled
## Problem:

When assign default values for property parameters and assign 'null' or 'undefined' value, resulting code generate instructions that do not make sense.

## Example:

source:
image

result:
image

## What is expected:

Probably no code at all as result since it does assign + check, but logically it is invalid

### LINKED ISSUE #1423: Wrong output with preceding statements in function parameter initializer
Example:

```typescript
let i = 1
function foo(x = i++){
    return x
}
foo()
```
This puts the preceding statements for `i++` outside of the function.

### PR DESCRIPTION
**Pull request by GlassBricks**
Closes #1422
Also fixes #1423

### CODE REVIEW

File: src/transformation/visitors/function.ts [hunk 1]
```diff
@@ -24,19 +24,26 @@ import { transformBindingPattern } from "./variable-declaration";
...
-): lua.Statement {
-    const parameterValue = value ? context.transformExpression(value) : undefined;
-    const assignment = lua.createAssignmentStatement(parameterName, parameterValue);
+): lua.Statement | undefined {
+    // const parameterValue = context.transformExpression(value);
+    const { precedingStatements, result: parameterValue } = transformInPrecedingStatementScope(context, () =>
+        context.transformExpression(value)
+    );
+    const statements = [...precedingStatements];
+    if (!lua.isNilLiteral(parameterValue)) {
```
Comment (Perryvw) [inline src/transformation/visitors/function.ts:35]: Couldn't we do that at the top of the function? Check ts.isUndefinedLiteral || ts.isNullLiteral? That way we can avoid a whole bunch of transformations.

File: test/unit/functions/functions.spec.ts [hunk 2]
```diff
@@ -107,6 +107,48 @@ test("Function default binding parameter maintains order", () => {
...
+            return x ?? "nil";
+        }
+        return [a.get("foo"), foo(), a.get("foo")];
+    `
+        .withLanguageExtensions()
+        .tap(builder => {
+            const lua = builder.getMainLuaCodeChunk();
+            expect(lua).not.toMatch("    x = nil");
+        })
+        .expectToEqual(["bar", "nil", "baz"]);
```
Comment (Perryvw) [inline test/unit/functions/functions.spec.ts:149]: Would prefer if you could just use `expectToMatchJsResult` here, using default TS (no language extensions) with an object and helper function for example.

**Review by Perryvw — CHANGES_REQUESTED** [review]

File: src/transformation/visitors/function.ts (hunk 1 above)
Comment (GlassBricks) [reply src/transformation/visitors/function.ts:35]: Doing this makes this the most general; it applies to anything that results in a nil literal (casts/parenthesis/etc, language extensions, plugin output) and shouldn't have any performance impact.

File: test/unit/functions/functions.spec.ts (hunk 2 above)
Comment (GlassBricks) [reply test/unit/functions/functions.spec.ts:149]: Could you give a more detailed example what that would that look like?
Besides using language extensions, I can't think of a good method that transpiles to nil (not just returns undefined) that works both in JS and lua.

Comment (Perryvw) [reply test/unit/functions/functions.spec.ts:149]: Oh I see what you are testing now, that seems maybe a bit too specific but okay

**Pull request merged**